### ISSUE TITLE
[FEATURE] Klipper support

### ISSUE THREAD

**Issue by leptun** [vintagepc/MK404#263]
@vintagepc I'd like to get Klipper working in the SIM as I need to do some comparison to PrusaFirmware. Only problem is that it currently doesn't quite work. The connection to the host software times out. I checked in uart_pty.cpp and data is received and sometimes even read by the code, but at some point the XOFF hook is engaged and the firmware gives up entirely. I'm quite unsure where to start looking at this issue. I suspect that there might be some unimplemented function that Klipper uses and Marlin/Prusa-Firmware doesn't, but I would expect some kind of warning for that. There might also be a deadlock in the firmware, but I don't know how to test for that. Before we had the "D" key for getting the PC, but now that option is gone and you have to use GDB (no idea how to do this) to get the position in the code where it locks up.

To get klipper running I followed the following tutorial: https://github.com/KevinOConnor/klipper/blob/master/docs/Installation.md
I also had to do some python2 trickery since I was using Ubuntu 20.04 and had to use some renamed packages.

MK404 run command: `./MK404 Prusa_MK3S -m -s -f ~/klipper/out/klipper.elf.hex`. `-n` didn't help.
Configuration file that I use on my MK3S: [printer.cfg.zip](https://github.com/vintagepc/MK404/files/5348626/printer.cfg.zip)

I hope that you will find some free time to help me with this issue/request

**Comment by vintagepc** [comment]
I can take a look at this - can you post the ELF you built? That will potentially save me some initial setup time figuring out how to build it.

I'll trade you - if you have some time to look at #253 and verify the Allegro behaviour against hardware behaviour for me it would be much appreciated and I can close out that too - I seem to remember you do have a MK2.

**Comment by vintagepc** [comment]
Also, I drafted this quick how-to on debugging AVR firmware with gdb:
https://github.com/vintagepc/MK404/wiki/Debugging-Printer-Firmware-with-MK404

**Comment by leptun** [comment]
Here you go. Not sure how helpful this will be since you also need the host software for it to work properly (and maybe octoprint as well if you actually want to test anything). Building it is quite easy especially since the host software and avr firmware are bundled together. Fot the AVR firmware, the default configuration in `make menuconfig` is ok. 
[klipper.elf.zip](https://github.com/vintagepc/MK404/files/5348772/klipper.elf.zip)

Eh... I do have a mostly stock MK2.5S on which I can do some testing, but it will take a bit till I have access to it since my high school got closed for 2 weeks because of corona. I can look at what went wrong though and maybe I'll find the issue.

I'll try to follow the Debugging instructions and maybe I'll find something interesting/useful

**Comment by vintagepc** [comment]
Looks like the Klipper FW is stuck in a WDR loop for me - If you run MK404 with `-vvv` you get endless
```
WATCHDOG: timer fired.
WATCHDOG: timer fired.
WATCHDOG: timer fired.
WATCHDOG: timer fired.
WATCHDOG: timer fired.
WATCHDOG: timer fired.
```

That might explain why it won't start... :)

**Comment by vintagepc** [comment]
I did some more digging... looks like klipper isn't responding properly. I don't see any "uart_pty_in_hook" log messages that indicate the AVR is sending serial data. Only incoming data from the host.

**Comment by leptun** [comment]
It’s not sending data back, but at least it’s reading it up to some point.
Doesn’t klipper actually use the wdt? That would be interesting.

**Comment by leptun** [comment]
@vintagepc Interesting detail that might help you: klipper actually uses also the ReadyToSend interrupt, not only the DataReceived interrupt. Maybe that one is broken.

**Comment by vintagepc** [comment]
unfortunately that may doom this from the start. 

I'm hoping it might be possible to just bypass that (since the PTY is essentially always able to receive data) but that won't work if klipper expects to be able to use that signal as a means of controlling the device. 

RTS/CTS are hardware control lines, and much like DTR cannot be simulated with a PTY.  We'd have to write a fake serial 
port device driver with those IOCTL calls, and that would immediately break our compatibility with anything but Linux. 

Another possibility would be to find a way to bind the sim to a real serial port and set up some sort of serial loopback connection (which would require two ports and physical hardware)

Option 3 is hacking up both projects to implement some sort of out-of-band flow control to mimic RTS/CTS.

**Comment by leptun** [comment]
Wait... I think I sent you on a wrong path. I accidentally used the Serial equivalent names instead of the actual interrupt vector names.
`USARTx_RX_vect` is used for receiving data.
`USARTx_UDRE_vect` is used for transmitting data.
Sorry for causing confusion. The atmega2560 doesn't even have RTS/CTS support. The only chip I know that supports this is the atmega32u4 and similar.

**Comment by vintagepc** [comment]
hmmm.. I may need to look at that in depth then. I know for the MK3 I had to implement a workaround because the UDRE bit was not cleared after changing the serial port config - and it would hang on boot. I wonder if there is a similar occurrence here and it gets stuck high too. 

The other thing I noticed is it expects the baud rate to  be much higher than default. It shouldn't cause any issues with PTYs, but could be that it's not handled correctly somewhere.

**Comment by leptun** [comment]
Klipper has a workaround for this UDRE issue:
```
// Tx interrupt - data can be written to serial.
ISR(USARTx_UDRE_vect)
{
    uint8_t data;
    int ret = serial_get_tx_byte(&data);
    if (ret)
        UCSRxB &= ~(1<<UDRIEx);
    else
        UDRx = data;
}
```
Btw, the MK3 firmware doesn't actually use the UDRE interrupt. Instead it just does the polling method and blocks the main loop

**Comment by vintagepc** [comment]
Makes sense - though whether we are polling or using the interrupt to wait for it to go low... the problem is the same if it never does :rofl:

**Comment by leptun** [comment]
[printer.cfg.zip](https://github.com/vintagepc/MK404/files/5355109/printer.cfg.zip)
Updated config that adds the ambient sensor on the einsy (according to latest push to master)

**Comment by leptun** [comment]
Just tried writing some basic arduino code with which I could test the serial.c from klipper.
I am disappointed to report back that it works as it should, so there might be something else that breaks the firmware.

**Comment by vintagepc** [comment]
yes, that's both good news and bad news. 

When I inspected it yesterday, the place the firmware was looping was in sched.c

```
                do {
                    irq_wait();
                } while (tasks_status != TS_REQUESTED);
```

**Comment by vintagepc** [comment]
I recall reading something about there being a built-in 2 instruction delay surrounding interrupt enable/disable.... The other thing that comes to mind is the way SimAVR terminates; there's a special case where if you sleep with interrupts disabled, it will exit the AVR run routine. So perhaps this is in some way falling afoul of that. 


The crash sounds like we hit an assertion failure and/or an over/underflow somewhere. But definitely a starting point for figuring out what is up. 

Sounds like there might be an opportunity for an upstream simavr patch out of this too.

**Comment by vintagepc** [comment]
found it
https://github.com/buserror/simavr/issues/376

**Comment by leptun** [comment]
Btw, simavr crashes after the connection is established. The LCD is initialized and the steppers are disabled and the fans spin a bit. I suspect that the simulator crashes while enabling some functionality of the MCU. The elf file provided is more or less generic. The firmware only knows it's pinout once the connection is established

**Comment by vintagepc** [comment]
:+1: 

the `-vvvv` argument may prove useful there; that will print some internal logging channels that can indicate whether it's on our end or a SimAVR thing.

**Comment by leptun** [comment]
[klippy.log](https://github.com/vintagepc/MK404/files/5355896/klippy.log)
Just for reference

**Comment by leptun** [comment]
@vintagepc How can I make the GL information stop being reported when using -vvvv? Even if I ignore the GL stuff, nothing extra is printed when simavr crashes

**Comment by vintagepc** [comment]
There's an if() in MK404.cpp that checks for the verbosity and calls GLDebugcallback to get notified on GL logging events. You can just comment that out temporarily.

**Comment by leptun** [comment]
![image](https://user-images.githubusercontent.com/17808203/95616328-4aa7ea00-0a72-11eb-8ebb-d0f11aa8a912.png)
I was not expecting that this will be an error in the LCD code. I'll try disabling the LCD in klipper and see if it fixes itself

**Comment by vintagepc** [comment]
That looks like an OOB on the vRAM on the LCD. 

`m_vRam[m_uiCursor] = ...` -> m_uiCursor seems to have been set to a value > 104.

**Comment by leptun** [comment]
And we have another issue...
```
Send: Status
Recv: // MCU 'mcu' shutdown: Rescheduled timer in the past
Recv: // This generally occurs when the micro-controller has been
Recv: // requested to step at a rate higher than it is capable of
Recv: // obtaining.
Recv: // Once the underlying issue is corrected, use the
Recv: // "FIRMWARE_RESTART" command to reset the firmware, reload the
Recv: // config, and restart the host software.
Recv: // Printer is shutdown
Recv: // Klipper state: Not ready
Recv: !! MCU 'mcu' shutdown: Rescheduled timer in the past
Recv: ok
```

**Comment by leptun** [comment]
Initializing code for timer1
```
// Activate timer dispatch as soon as possible
void timer_kick(void)
{
    timer_set(timer_get() + 50);
    TIFR1 = 1<<OCF1A;
}

static inline void timer_repeat_set(uint16_t next)
{
    // Timer1B is used to limit the number of timers run from a timer1A irq
    OCR1B = next;
    // This is "TIFR1 = 1<<OCF1B" - gcc handles that poorly, so it's hand coded
    uint8_t dummy;
    asm volatile("ldi %0, %2\n    out %1, %0"
                 : "=d"(dummy) : "i"(&TIFR1 - 0x20), "i"(1<<OCF1B));
}

static inline uint16_t timer_get(void)
{
    return TCNT1;
}

void timer_init(void)
{
    irqstatus_t flag = irq_save();
    // no outputs
    TCCR1A = 0;
    // Normal Mode
    TCCR1B = 1<<CS10;
    // Setup for first irq
    TCNT1 = 0;
    timer_kick();
    timer_repeat_set(timer_get() + 50);
    timer_reset();
    TIFR1 = 1<<TOV1;
    // enable interrupt
    TIMSK1 = 1<<OCIE1A;
    irq_restore(flag);
}
```

I'm actually a bit confused why they don't use the CTC mode, but I digress.

**Comment by leptun** [comment]
Aaaand the cursor position where it attempts to write is 117 decimal. So outside the screen. The commands it sends to the LCD are very suspicious. Wondering if the timer issue is linked to this (no idea how the lcd is implemented in klipper).
This looks suspicious:
```
ProcessWrite command 0write when still BUSY
```

**Comment by vintagepc** [comment]
IIRC that's just a warning that it's writing to the display faster than the real thing can handle. 

Does klipper read back data from the LCD at any point? it might not be wired up to support that as the factory firmware only ever writes.

**Comment by leptun** [comment]
RW line is not implemented on the einsy/ultralcd, so it shouldn’t ever try something like this.

**Comment by vintagepc** [comment]
Yes, something seems very wrong with the LCD config or I/O... I tested it by expanding the ram to 128 and got this:
![image](https://user-images.githubusercontent.com/53943260/95629642-58e40f00-0a4e-11eb-8708-0dfe3479e751.png)

**Comment by vintagepc** [comment]
It almost looks like all the LCD bytes are off by a nibble. 

I wonder if it's worth opening a request for help on their github; at the very least they might be able to interpret the klipper logs for us to help pinpoint the issue.

**Comment by vintagepc** [comment]
I should note that I recompiled locally and tweaked the irq_wait() function ASM line to have two `nop` calls. Now it behaves similarly to the one you provided - but they both still sometimes get in a state where it won't seem to connect at all.

**Comment by leptun** [comment]
The LCD issue is not really related to the timer in the past issue as even if the LCD is disabled in the config, the timer still executes in the past.

I also noticed that it fails to connect sometimes. When that happens I just restart the klipper service and also the printer and then it “works”

**Comment by vintagepc** [comment]
Yes - I don't think the display itself is the cause - I think it is just a sign that something is going wrong earlier in the startup and causing problems downstream.

**Comment by vintagepc** [comment]
Did a bit more poking, have the following notes to add:
 - changing the SEI code to -1 instead of -2 does indeed appear to resolve the hang; it boots as expected with the stock build then. 

- I can confirm the problem appears to be the combo of klipper and simavr. I modified the simduino example on a stock simavr checkout and it exhibits the same "timer scheduled in past" issue.

**Comment by vintagepc** [comment]
Made some more progress today. Display issue is due to bad init and our virtual display not tossing out writes while busy. 

I hacked around the timer stuff temporarily and need to dive in more. 

![image](https://user-images.githubusercontent.com/53943260/95697367-7fe83f80-0c0c-11eb-9d35-f89984c61486.png)

**Comment by leptun** [comment]
Did you push these changes anywhere?

**Comment by vintagepc** [comment]
Just pushed to 263-klipper-support. git submodule update required for simavr tweak. 

I temporarily worked around the timer in past by commenting it out in the klipper source until I can dive in deeper.

**Comment by vintagepc** [comment]
Just pushed up a fix for the initial bootup issue (submodule update required). 

We get further now but attempting to do anything results in "timer too close", so that is the next problem to debug.

**Comment by leptun** [comment]
Issues left to fix:
- [ ] Bed heater is always enabled. Probably need to revert some of the hacks from PrusaFirmware with the timers
- [ ] Axes hit the endstops immediately. This is not consistent as sometimes the Z will try to home, otherwise it homes instantly. X/Y always home instantly.
- [ ] Fans seem to work, but we get the timer too close error sometimes.
- [ ] Hotend heater seems to work, but the timer issue didn't let me pass ~40C
- [ ] `Timer too close` error

**Comment by vintagepc** [comment]
Knocked off the first one; `-n` now disables the bed softpwm hack that causes this.

**Comment by vintagepc** [comment]
Hmm... I'm seeing this in the logs. Might have to artificially slow down the sim a bit...
`freq=41679345`

**Comment by vintagepc** [comment]
hot damn, that seems to have worked!

Looks like the endstops need to be inverted in the printer config, they're tagged with a `!` for active low but DIAG is active high as we use it with the official firmware.Fixing that causes X/Y to home as expected.

Not sure what's up with Z; that doesn't seem happy. Digging in. 

It now throws a "not heating at expected rate" error - good sign, means I think we're close to just fine-tuning the config!
![image](https://user-images.githubusercontent.com/53943260/95868077-ddc57600-0d37-11eb-8069-35d3c00fe13a.png)

Pushed up the speed hack - in Board.c RunAVR():
```

		if (m_pAVR->cycle > cnt+2000)
		{
			usleep(10);
			cnt = m_pAVR->cycle;
		}

```

it's likely machine specific, for now just play with the cnt+ [value] to find something that klippy.py reports `freq=` as close to 16MHz. Longer term I'll look at a way of constraining that dynamically and/or figuring out why this firmware clocks at 41MHz whereas the stock one seems to be more true-to-life. Could very much just be the emulation load.

**Comment by leptun** [comment]
@vintagepc Please check the config. I think they are using the other DIAG pin mode of functioning where it's active low only, not push-pull

**Comment by vintagepc** [comment]
Ah, ok - in that case we'll need to implement the config register in the TMC2130. I forgot about that possibility and will implement it. In any case it's a minor tweak that can get you underway while I polish things :+1: 

I'm sure @KevinOConnor will be interested in the progress here :smile:

**Comment by vintagepc** [comment]
Pushed up a proper diag_pushpull implementation.

**Comment by KevinOConnor** [comment]
> Longer term I'll look at a way of constraining that dynamically and/or figuring out why this firmware clocks at 41MHz whereas the stock one seems to be more true-to-life.

FWIW, I implemented the code at: https://github.com/KevinOConnor/klipper/blob/cef9cc29b8f150681cd24ca6f051fc9f03fc924b/scripts/avrsim.py#L102 to do something similar for simulavr.  (Basic idea is to alter the sleep amount based on how much the clock rate differs from the system time, while also preventing an unfortunate "slow emulation" from incurring a future "fast emulation to catch up" - as that would just make things worse.)

-Kevin

**Comment by vintagepc** [comment]
> > Longer term I'll look at a way of constraining that dynamically and/or figuring out why this firmware clocks at 41MHz whereas the stock one seems to be more true-to-life.
> 
> FWIW, I implemented the code at: https://github.com/KevinOConnor/klipper/blob/cef9cc29b8f150681cd24ca6f051fc9f03fc924b/scripts/avrsim.py#L102 to do something similar for simulavr. (Basic idea is to alter the sleep amount based on how much the clock rate differs from the system time, while also preventing an unfortunate "slow emulation" from incurring a future "fast emulation to catch up" - as that would just make things worse.)
> 
> -Kevin

Yes - simAVR has similar code internally to avoid sim and wall time divergence. I do not know why it doesn't seem to be working here (for either firmware - stock f/w also seems to be clocking much faster).

**Comment by vintagepc** [comment]
Simulator's PID parameters:

// PID parameters: pid_Kp=63.285 pid_Ki=12.785 pid_Kd=78.316

I know it seems to stall out as it gets close to the target with the defaults.

**Comment by vintagepc** [comment]
Teaser... Still fragile but I managed to get it to home, heat up, and run a VFA print:

The fuzzy section towards the bottom is with X nonlinearity on, the rest is with it off. 

![image](https://user-images.githubusercontent.com/53943260/96022708-46d3e900-0e1f-11eb-923a-ae2091ce4bfa.png)

What's very interesting... is the artifacts do not line up vertically. This would be *very* interesting to compare with a real print.

![image](https://user-images.githubusercontent.com/53943260/96022982-b944c900-0e1f-11eb-822c-20f06f8291b6.png)

**Comment by wavexx** [comment]
mmh, there's always the E axis involved right? (to get the width).

**Comment by vintagepc** [comment]
ooooh, true; if the E behaviour is very different it would change where the excess and under-extrusion happens

**Comment by wavexx** [comment]
I mean, XYZ cannot drift, so the only motor that can shift here is E (unless we have a bug somewhere of course ;))

**Comment by vintagepc** [comment]
A bug is always possible but would be very strange given the way the positioning works.

**Comment by wavexx** [comment]
Looks about right to me BTW. The different level of artifacts might suggest the klipper config has a different e-step/mm spacing perhaps?

**Comment by vintagepc** [comment]
I just pushed up some code with `--skew-correct` which attempts to correct clock skew in the fast direction (it can't help if the system is too slow, of course). 

It seems to make the frequency a lot more stable which in turn stabilizes Klippy.

I'm attempting to run a print with pronterface now to see if it is effective at preventing "timer too soon"s.

**Comment by DRracer** [comment]
@vintagepc was your print aligned with X or Y axis? If not that might be the cause of those non-aligned artefacts. We can see them on our FW as well, but so far they have been perfectly aligned. Another possibility is MBL kicking in

**Comment by vintagepc** [comment]
Yes - this was my X-only single-wall VFA test gcode that I have checked in.

**Comment by vintagepc** [comment]
@leptun I reworked the skew correction a bit and on my machine the Klipper freq parameter is now very close to 16 MHz. I have not yet seen a "timer too close" after this change. I do get a "timer in the past" but it happens at the exact same point in the gcode so I suspect it is badly sliced with too high a travel speed. 

Also, pushing the pause key will now print out stats, e.g.:

```
Sim is behind by164us!
Pause: 
Sim is ahead by2966us!
```

A few things that may help with your VM speed, too:
- Make sure you run cmake with `-DCMAKE_BUILD_TYPE=Release`. That enables compiler optimizations and removes debugging symbols from the code.
- If you haven't already, double check your VM has at least three virtual CPUs, if not more, since there are at least three threads running  - UI, AVR core, and serial.

Let me know if this fixes the issues for you too.

**Comment by leptun** [comment]
I’ll try it later today. 👍

**Comment by leptun** [comment]
I just pushed a small patch to a mistake (probably forgot to stage this change). Switching to Release did the trick for me in the VM. Now with the speed patch I get 33Mhz so I actually need to slow down :). When pausing I am waaaay ahead. I assigned 4 virtual CPUs to the VM from the beginning. 4GB of RAM is probably more than enough especially when not rendering extrusion. Is the current code supposed to ensure 16MHz regardless of (excess) system performance? Because if yes, then it is not working for me. Not yet anyway

**Comment by leptun** [comment]
I tracked the issue. Forgot to add `--skew-correct` when running MK404. Now it works perfectly. Just a bit under 16Mhz, which is fine. Thank you again for implementing support for klipper. Now I wonder, how are we going to make this part of the main program and not just a list of hacks to enable for klipper? I'd guess that we could have something like `--fw <PrusaFirmware/klipper/Marlin/generic(aka. nohacks)>` so that we don't have to remember what hacks to enable every time. If no `--fw` is specified, then we use the standard hacks for PrusaFirmware, or we use no hacks at all. The decision is yours to make.

**Comment by vintagepc** [comment]
I've been pondering something similar; I don't see any reason we couldn't do that if I can find a way to poke the incoming arguments, e.g. if `--fw klipper` is set it automatically enables `--skew-correct` and `-n`. Should be fairly easy to maintain.

Glad to hear it's working for you! I'll review/polish the changes we made in the klipper branch and work at merging it up to master.

**Comment by vintagepc** [comment]
Alas TCLAP has no nice way to handle convenience args; if we start needing more of them I'll look at forking the project and adding some sort of support to deal with them dynamically. 

But for now I've just added --klipper and --marlin as shorthand for their respective -n and skew flags.

**Comment by KevinOConnor** [comment]
What was the verdict on the "fine artifacts" result reported at: https://github.com/vintagepc/MK404/issues/263#issuecomment-708543765 .  I'm not familiar enough with the test to understand if the results were something with Klipper or something with the simulation.

-Kevin

**Comment by vintagepc** [comment]
We don't know for sure yet - investigating that stepping behaviour as compared to other firmware is part of the reason for this request to support Klipper. There's some more discussion on the details of the visualization in #114; The image is essentially turning the stepper steps into something we can easily interpret visually - it doesn't mean Klipper is doing anything "wrong" - just that it's different from what we've seen with stock firmware and we'll have to look a little deeper to understand it. :+1:

**Comment by leptun** [comment]
This is what the VFA print looks like when printed using Klipper on a MK3S. The following 3 pictures are made using different flow rates (M221). Not sure if this diagonal pattern we are seeing is what was generated in MK404 or if it's some other issue (most likely TMC2130 related).
![Image from iOS (13)](https://user-images.githubusercontent.com/17808203/96346900-f53d8f80-10a6-11eb-9496-6f9ecfdbe33c.jpg)
![Image from iOS (14)](https://user-images.githubusercontent.com/17808203/96346903-f8d11680-10a6-11eb-9232-599fa4305d7e.jpg)
![Image from iOS (15)](https://user-images.githubusercontent.com/17808203/96346906-fb337080-10a6-11eb-9ba0-952d9b6416f5.jpg)
@vintagepc What settings did you use for the picture above with klipper diagonal lines?

**Comment by vintagepc** [comment]
I did not do anything besides connect, load the file in Pronterface, and print (so default speeds etc as given in the gcode). 

The fuzzy section in the middle was purely toggling X nonlinearity via the context menu. 

Your pictures are very interesting; traditionally the diagonal lines are E nonlinearity since the E position is what is changing from layer to layer. 

The uneven end is actually very interesting - I saw that in the simulation too but I chalked that up to LA.

**Comment by vintagepc** [comment]
Just saw the edit - I bet if we ran a simulation with E nonlinearity you would see the same periodic underextrusion.

**Comment by leptun** [comment]
The uneven end is caused most likely by poor adhesion on my printer. With higher flow rate I managed to make the parts stick more to the bed, but it's still not great. This really needs a brim for real prints.

**Comment by leptun** [comment]
@vintagepc Actually, how uneven was the end of the part in MK404? I might have some minor issues with LA on actual prints with Klipper at very high speed and maybe this is related.

**Comment by KevinOConnor** [comment]
I'm not all that familiar with this topic, so it's hard for me to comment.  (VFA is "very fine artifact"? What's "LA"?)

I can say that the order of Klipper step pulses is unlikely to match Marlin/Prusa.  Those firmware use the Bresenham algorithm which causes strict (and incorrect) ordering between step pulses.  Klipper calculates and schedules a precise time for every step pulse irrespective of the step pulses occurring on other stepper motors.  So, there wont be a clear ordering of step pulses between steppers - except as dictated by toolhead movement.  That said, Klipper does allow the final step times to move within a window of 25us to improve compression (think +-12.5us).  So, in summary, if the simulation is dependent on step pulse ordering than that may show up differently with Klipper.

As for non-linearity - I'm guessing you are referring to the fact that stepper motor coils may not be perfectly balanced, and are also referring to the tmc2130's ability to alter the sine wave to account for that?  If so, Klipper doesn't currently support that.  Though I understand that @Arksine has implemented that in one of his branches.  If I understand correctly, though, he didn't find much of an improvement with it on his printers.

-Kevin

**Comment by vintagepc** [comment]
Yes - VFA = "vertical fine artifacts" - fine striations in the print surface that line up vertically. LA = linear advance. 

We are indeed simulating the nonlinearity of the stepper coils for the goal of understanding its effects. Thanks for your insight on the stepping; it does explain some of what we are seeing; the visualization of MK404 is limited to the resolution of the E steps (since that is what we use to do the drawing). So if E does not "tick" consistently relative to X and Y ticks,  then it could make sense that the simulated VFAs do not line up at this time (but I suspect they would if we re-ran the simulation at a higher level of E microsteps.)

@leptun - I reran that print, more pictures:
Starting edge:
![image](https://user-images.githubusercontent.com/53943260/96347862-9a8a4080-1072-11eb-8cc1-46b072429a25.png)

The big gashes are E-nonlinearity enabled - that looks a lot like your 95% picture!
![image](https://user-images.githubusercontent.com/53943260/96347891-dcb38200-1072-11eb-9223-154ddee964f7.png)

Here's what the right edge of my print looks like:

![image](https://user-images.githubusercontent.com/53943260/96347937-32882a00-1073-11eb-8146-083165c1706b.png)

Closeup of the nonlinear X section:
![image](https://user-images.githubusercontent.com/53943260/96347959-66fbe600-1073-11eb-9f80-922d2a768620.png)

**Comment by leptun** [comment]
LA is Linear Advance, I think it is called Pressure Advance in Klipper.
I actually have no idea right now if it's stepper inconsistency that is causing the lack of material or if it's some other mechanical issue (or maybe something related to the drivers).
The start of line wobble issue might be because of what you said above with VFA: Nothing will be drown unless E does a step. As a result, the start of the line might shift with up to a step in length every layer because of this. I think you can actually observe a pattern in that as well. So I don't think this has anything to do with LA.

**Comment by KevinOConnor** [comment]
Okay, just so I understand, every time there is an E step pulse, the visualization draws a line from the XY position of the last E step pulse to the XY position at the current step pulse?

> Here's what the right edge of my print looks like:

I suspect that is an artifact of the simulation - I'd guess there is not a even number of E steps per linear movement.

To explain with an example, consider a line segment that requests 200 X steps with 87.654 E steps over 1s.  With the Bresenham algo, you'd get 200 X steps every 5ms, and 87 E steps sometimes spaced 10ms and sometimes spaced 15ms apart.  In contrast, Klipper would produce 200 X steps every 5ms and an E step every ~11.408492ms apart.  That is, Klipper produces both smooth E movement and it can effectively produce "fractional e-steps" by using proper tuning.  The timing of the E step pulse at the start of the next extrusion move is dependent on the fractional e-steps position.

So, in summary, I suspect you're seeing uneven corners on Klipper because the simulation is getting confused by the timing (wrt fractional e-steps) of the first step in the following movement.

> LA is Linear Advance, I think it is called Pressure Advance in Klipper.

Ah, okay.  They have similar concepts, but the implementation is notably different.  (Klipper uses [smooth pressure advance](https://www.klipper3d.org/Kinematics.html#pressure-advance) which does not introduce instantaneous velocity changes.)  It does need to be calibrated using the method described at https://www.klipper3d.org/Pressure_Advance.html before using it.

> I might have some minor issues with LA on actual prints with Klipper at very high speed and maybe this is related.

FWIW, though, your image doesn't look like it is suffering from pressure advance issues to me.

-Kevin

**Comment by vintagepc** [comment]
>Okay, just so I understand, every time there is an E step pulse, the visualization draws a line from the XY position of the last E step pulse to the XY position at the current step pulse?

Yes, at least for the `_HR` modes which is what we're using to see the stepper artifacts. Notably, it also does not handle "coasting" type operations that compensate for trailing ooze on extrusion moves - it will also render gaps there until we get around to #265

**Comment by 3d-gussner** [comment]
I have been able to print ONE simple print with MK404 and Klipper, the rest just fails.

Here my MK404 Klipper configs:

- gcodes
- finished print ply file
- failed print ply file
- Klipper "printer.cfg" from @leptun
- PrusaSlicer 2.2 Klipper config
- Klipper hex file
- Klippy.log files

I got one print finished after:
- Deleting Prusa_MK3S_eeprom/flas/xflash.bin files in MK404 (not sure if that helped)
- restarting Klipper/Octoprint service before every print
- Minimizing MK404 both windows
- Do nothing else on computer

Klipper version commit 2bcf06a295e5c9696b1bf556a0520b49957a369f
MK404 version commit 208ef256e27f010a5868fc1651f923729334ef32

@KevinOConnor Do you have any idea what I can do to get it working?
@vintagepc As my PC isn't the most performant one and a 2h real time 3D Benchy takes 18+h to finish in MK404, do you have any idea how we could improve the performance?


[Uploading MK404-Klipper.zip…]()

**Comment by 3d-gussner** [comment]
Short update: Even without `-g lite` Klipper crashes after short time on the 3D Benchy print.

**Comment by vintagepc** [comment]
Looks like your files did not attach, the link is invalid. 

Was MK404 built in release mode? (`-DCMAKE_BUILD_TYPE=Release`)? It is much more performant with that.

**Comment by 3d-gussner** [comment]
Upload fixed.
I could print a 3D Benchy up to Z2.80mm and it crashed again.

**Comment by vintagepc** [comment]
Hmm... "Move queue empty" is indeed a new one. I'm not sure what would cause that, since there isn't a serial bottleneck like there is on real hardware... clock frequency looks good and should be stable enough for klipper to not hit timer-related errors. 

I'm not an expert in Klipper logs but it almost looks like it's having issues reading the source .gcode file on the klipper host.

**Comment by KevinOConnor** [comment]
A "Move queue empty" error indicates the host was not able to make accurate predictions of the mcu clock.  (Specifically, the host is tasked with queuing movement only when the mcu has space for it - if the host can't predict when the queue entries will become free it will result in a "move queue empty" error.)

-Kevin

**Comment by vintagepc** [comment]
@KevinOConnor Feels like we are close. 

I managed to get lucky and get 30% of a benchy before it failed with "lost communication". However, I suspect that was due to hitting the preallocated vector size and the simulator needing to allocate more space. (The fact we can get that far, yet sometimes it fails right on the first layer leads me to think we may be riding right on the edge of sufficient timing synchronization.)

Forgive me if this is a stupid question in the context of how Klipper operates - but is it possible to have it queue up a few more entries in advance with the assumption they will happen at exactly the predicted time and reduce the worry about divergence from the expected timing?

That combined with some improvements to the skew correction I am working on may be sufficient to help it smooth over the bumps. Or - if you have other ideas for liberties we could take since it is a simulated environment then I'm all ears. An out-of-band feedback loop of some kind might also work.

Ugly as it is I've stopped trying to sleep to match simulation time to wall time; I'm getting a more stable clock by spinlocking and rechecking the clock until we meet or exceed the wall time. 

For @3d-gussner since you were trying to get a Benchy in HRQ mode:

[Export.ply.zip](https://github.com/vintagepc/MK404/files/5470004/Export.ply.zip)

**Comment by KevinOConnor** [comment]
> is it possible to have it queue up a few more entries in advance with the assumption they will happen at exactly the predicted time and reduce the worry about divergence from the expected timing?

The "move queue empty" error is a bit misleading - it probably should report "move queue full".  The error indicates that the host tried to queue another move on the micro-controller, but there was no available memory to store it.  The host is tasked with only queuing moves when there is space for them (by tracking when old moves complete).

As to your question though, could the host not use as much mcu memory so it doesn't hit the limit?  The host code does try to give sufficient extra time to avoid that.  The code is at https://github.com/KevinOConnor/klipper/blob/cef9cc29b8f150681cd24ca6f051fc9f03fc924b/klippy/clocksync.py#L115 - it adds TRANSMIT_EXTRA (.001s) plus three times the measured standard deviation of the frequency.

FWIW, in my previous tests with simulavr, I didn't find tweaking that extra time helped with stability.

> That combined with some improvements to the skew correction I am working on may be sufficient to help it smooth over the bumps. Or - if you have other ideas for liberties we could take since it is a simulated environment then I'm all ears. An out-of-band feedback loop of some kind might also work.

Can you describe the algorithm you are using to pace the clock?

> Ugly as it is I've stopped trying to sleep to match simulation time to wall time; I'm getting a more stable clock by spinlocking and rechecking the clock until we meet or exceed the wall time.

FWIW, I got around that in my simulavr code by using an offset in the usleep time - that is, usleep() will always sleep a little more than requested, but it's harmless to always sleep slightly less to account for that: https://github.com/KevinOConnor/klipper/blob/cef9cc29b8f150681cd24ca6f051fc9f03fc924b/scripts/avrsim.py#L102

Separately, does simavr use accurate timing for serial port reads?  That is, on a 16Mhz avr, will it take at least 640 cycles for each byte to be read and transmitted?  If not, that could also be messing up Klipper's timing.

-Kevin

**Comment by vintagepc** [comment]
OK, so that means we're not finishing things as fast as we should be - as opposed to running out of "things to do"

>Can you describe the algorithm you are using to pace the clock?

All I've been experimenting with are variants of checking the "simulation time" in ns against the "wall time" and then doing *something* to make up the difference at regular intervals. The size of those intervals and specifically what I am doing to idle are the knobs I've been twiddling. I'm using the graph script to plot Klippy's calculated MCU frequency; this is one of the better ones I've managed to get but clearly it's still too unpredictable. I'm guessing with a proper micro this line is very flat. 

![6](https://user-images.githubusercontent.com/53943260/97809141-356d4800-1c39-11eb-9377-4809dc7e9951.png)

Many look more like this, where it dips and recovers after the virtual print stops. 

![5000](https://user-images.githubusercontent.com/53943260/97809174-69486d80-1c39-11eb-8f67-7a8db4e0d26c.png)

>Separately, does simavr use accurate timing for serial port reads? That is, on a 16Mhz avr with 250000 baud, will it take at least 640 cycles for each byte to be read and transmitted? If not, that could also be messing up Klipper's timing.

Good thought - I know it will clock in data bytes at a time but I will have to look into that. We've definitely encountered some other peripheral I/F issues like that before (with I2C and SPI) given the nature of the simulation we're doing.

**Comment by vintagepc** [comment]
Hmm... I wonder if we are chasing the wrong underlying cause. I just managed to get a frequency graph that looks like this and it still failed with the queue error. 

![13](https://user-images.githubusercontent.com/53943260/97813336-77a48280-1c55-11eb-829f-261554577015.png)

**Comment by KevinOConnor** [comment]
FYI, I don't think the charts (or the data in the log in general) will help, as the frequency is only logged once per second.  What's going to cause problems is sub-second bursts and sub-second slow downs.  So, for example, if the simulation runs 20ms of simulation in 35ms, but then runs the next 20ms of simulation in 5ms it will show up fine in the graphs, but there's a very good chance it would cause problems internally in Klipper.

> All I've been experimenting with are variants of checking the "simulation time" in ns against the "wall time" and then doing something to make up the difference at regular intervals.  The size of those intervals and specifically what I am doing to idle are the knobs I've been twiddling.

FWIW, I didn't have luck with that approach in simulavr - the problem I had was that after a "slow down" the pacing would try to "catch up" and that would ultimately cause more timing havoc.  The scheme I finally came up with ( https://github.com/KevinOConnor/klipper/blob/cef9cc29b8f150681cd24ca6f051fc9f03fc924b/scripts/avrsim.py#L102 ) was to check pacing every 100us and to reset timing every 5ms.  That is, I allowed the simulation to "slow down" and "catch up" every 100us, but never allow it to "catch up" more than a couple of milliseconds.  If the simulation fell behind by a couple of milliseconds then that time is lost (the host can usually adopt).

I did also have to run the simulation very slowly (at 30% normal rate).  I suspect simavr is much faster than simulavr so that may not be an issue.  I would verify that the simulation thread isn't using any more than about 75% of a cpu though.

Also, you may want to wait 30 seconds after startup before starting a print - just so the host has a good sync on the clock.  (I suspect that wont help though.)

-Kevin

**Comment by vintagepc** [comment]
Hmm, that might explain why I am having more luck with my current approach, shy of checking the clock every simulated cycle. Still variable but I did get 80% of the way through a benchy once.

Rather than adding delay time every x cycles, this approach takes advantage of the fact that checking the nanosecond clock is relatively expensive (on the order of a few hundred ns). Then, depending on whether we are running ahead or behind, it increases or decreases the frequency of these checks one cycle count at a time. I think this should work as it smears the correction over time and does not result in significant jumps but a smoothed correction. 

I haven't checked where it ends up settling, but with prior attempts I was certainly attempting to correct for the time on roughly a 500 cycle count scale (about 31 microseconds), and in some cases even more often. 

One thing I keep forgetting to look at is the SimAVR cycle ticking; I don't think it "visits" every cycle count and if the CPU has small idle periods it will jump over those. I need to get a good picture of how big those jumps are.

**Comment by vintagepc** [comment]
It also occurs to me that SimAVR supports external clocking of AVR timers.  Perhaps we can make use of that to clock the simulation at a more regular rate.

**Comment by vintagepc** [comment]
Not finished yet but I pushed up the branch that let me get 80% of the way through a benchy, in case it helps with stability for others. If attempting large prints you'll likely want to adjust the preallocation in GLPrint.h as when we hit that limit Klipper will likely choke because the MCU does not respond in time. 

https://github.com/vintagepc/MK404/commits/263-Klipper-take-2

**Comment by vintagepc** [comment]
OK - I took a look at the timing on a microscale level - this is a plot of the "wall time (ns)" per cycle as time progresses with the current scaling algorithm. 

This sets me up to get some more useful information; now I can visually see how we are doing over time. In a perfect situation this chart should be a flat line at 62.5. 

![image](https://user-images.githubusercontent.com/53943260/98317839-486e7800-1fab-11eb-9fdd-aa79ecb1ec70.png)

So we can see some major excursions that are sporadic, and also that my current skew correction is complete crap and way too slow to respond :rofl: So the 80% benchy is a complete fluke.

**Comment by KevinOConnor** [comment]
FWIW, calling `clock_gettime(CLOCK_MONOTONIC_RAW)` is higher overhead than a regular `gettimeofday()` or `clock_gettime(CLOCK_MONOTONIC)` call (because on most machines the latter use the Linux vDSO).  You shouldn't need microsecond clock accuracy (but will need millisecond accuracy).  So, the normal Linux clocks should work okay.

-Kevin

**Comment by vintagepc** [comment]
@KevinOConnor So you're saying that it should be able to handle small excursions so long as we're tracking accurately somewhere around the millisecond scale? I am currently working at about a microsecond scale to hopefully get a better picture of what's going on - but once I understand the issue in depth I should be able to pull back a bit and operate on a larger timescale. 

I think part of it is there are ocassional excursions that are what are biting us; so I am currently trying to see if I can identify what is causing those - the bulk of the time the simulation seems to be pretty good at matching pace with the wall clock.

**Comment by KevinOConnor** [comment]
> So you're saying that it should be able to handle small excursions so long as we're tracking accurately somewhere around the millisecond scale?

Yes.

The current Klipper code measures the clock frequency by querying the current clock value once per second.  (The resulting offset/frequency could be skewed by comms jitter and the frequency can change over time, so Klipper performs a bunch of "linear regression" math to obtain better estimates from the raw measurements.)  Klipper can eventually get to an estimate with micro-second accuracy, but it isn't necessary.

> I think part of it is there are ocassional excursions that are what are biting us

I agree.  There are two main failure cases: the clock estimate is too low, or the clock estimate is too high.

For the too low estimate, the mcu may "race ahead" and not have necessary data buffered by the time it needs it.  This problem is often seen as a "Timer too close" error.  It's unlikely to strike here though, because Klipper tries to buffer a minimum of 100ms (and up to 2s) in the micro-controller.  So, this problem should really only occur it the host scheduling falls behind and mcu races ahead for a total of 100ms.

The other error is due to the micro-controller going too slow.  In this case, Klipper may try to buffer content in the mcu that can't fit because the mcu was too slow to process the previous content (often seen as a "move queue empty" error).  To avoid this case, Klipper uses TRANSMIT_EXTRA (in klippy/clocksync.py) along with `3*measured_clock_jitter` to avoid that case.

I suspect that Klipper is observing a pretty steady clock rate over one second intervals (and thus `measured_clock_jitter` is pretty low).  However, if a 7ms micro-slow-down happens to occur when Klipper is buffering some content then it'll trip up and throw an error.

You could always try increasing TRANSMIT_EXTRA from 1ms to 10ms if you think your micro-slow-downs are only in the ~7ms range.  There shouldn't be any harm in doing that.

Cheers,
-Kevin

**Comment by vintagepc** [comment]
Status update:  I've added more logging code and can now see that we are indeed losing handfuls of ms at random intervals. It's not consistent, and there are a few culprits. usleep/nanosleep  is a particularly bad offender since it will sporadically sleep 10x or more what was asked. Mutexes from the various cross-thread operations are another. I also suspect disk I/O is hurting things too as it might be blocking the files used as PTYs. 

In any case I'm making adjustments and adjusting TRANSMIT_EXTRA,  combined with the removal of some sleep calls and display mutexes, have so far gotten one complete benchy (second one is currently 64% after cleanup). However, this is without the full 3d visualizations. Those definitely have some mutexes and bottlenecks so I will look at whether I can address those next. If not it might be possible to make a lightweight version that tracks the model being printed but does not actually render anything to screen so that it does not have to worry about cross thread memory access.

**Comment by vintagepc** [comment]
Victory! I think thinks are in good shape now. May still find a few lingering issues but I've just completed two benchys without issue. 
![image](https://user-images.githubusercontent.com/53943260/98455182-3f141580-213b-11eb-9f72-03673b28be85.png)

**Comment by KevinOConnor** [comment]
Nice.  Out of curiosity, did you also end up changing TRANSMIT_EXTRA in Klipper?  And if so, what did you ultimately set it to?

-Kevin

**Comment by vintagepc** [comment]
Yes - I did end up changing it to 10ms as suggested and that seems to have done the trick to help with smaller irregularities (after I got rid of the larger ones caused by the sleeps and cross thread mutexes)

It still chokes on underpowered computers (we know that >4 cores/threads are required - since MK404 takes at least 4 threads by itself, not accounting for the Klipper process and whatever you are using to actually run the print). We had some issues on @3d-gussner's machine initially but observed it would print just fine if the 3d graphics were disabled.

In the end I successfully completed three benchy prints, one of which was even at 200% feedrate.

**Comment by KevinOConnor** [comment]
> I did end up changing it to 10ms as suggested and that seems to have done the trick to help with smaller irregularities 

Ah, makes sense.  I was thinking of changing the default in Klipper from 1ms to 5ms.  I'm a little leery of going to 10ms for the default, just because I'm afraid of introducing a regression (and I've not seen the problem outside of simulators).  I'm not sure if 5ms would help here though.

> It still chokes on underpowered computers

FWIW, it's fine to simulate a slower speed (eg, pace to 8mhz on a nominal 16mhz chip).  Klipper should produce identical results (other than the simulation taking longer) as Klipper times actions based on the main micro-controller clock.  Klipper doesn't really care about the actual clock speed - it just wants a clock it can accurately predict.

Cheers,
-Kevin

**Comment by vintagepc** [comment]
It might be sufficient for the simulation to use 5, I added logging as a feature and it will report "lost" time; usually I only see <5ms with any regularity - but I can definitely make it have significant losses by doing things that load the CPU cores or disk. 

I think the losses aren't necessarily the simulation speed but the context swapping and concurrency of other processes if your system doesn't have a dedicated core for each thread that isn't idle. At least in this case, we are spinlocking in order to get accurate time correction - which means that even if you run the simulated AVR slower it will still use the same amount of CPU time. 

Certainly it's already very impressive it works as well as it does at full speed. 

BTW - if there are features I can add that would make the simulator more useful to you don't hesitate to let me know. I appreciate your involvement in helping to get it running for our own purposes so I'm happy to reciprocate that if I can make it useful for you too.

**Comment by KevinOConnor** [comment]
Thanks.  It does seem like a useful tool.  Unfortunately, I haven't gotten a chance to spend time working with MK404, and I probably wont get much time in the short-term.

I do appreciate the offer and your efforts!
-Kevin
